sbin/chown/chown.8
.\" Copyright (c) 1990, 1991, 1993, 1994, 2003
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     from: @(#)chown.8	8.3 (Berkeley) 3/31/94
.\"	$NetBSD: chown.8,v 1.14 2023/05/20 00:54:15 pgoyette Exp $
.\"
.Dd May 19, 2023
.Dt CHOWN 8
.Os
.Sh NAME
.Nm chown
.Nd change file owner and group
.Sh SYNOPSIS
.Nm
.Oo
.Fl R
.Op Fl H | Fl L | Fl P
.Oc
.Op Fl dfhv
.Ar owner Ns Op Ar :group
.Ar
.Nm
.Oo
.Fl R
.Op Fl H | Fl L | Fl P
.Oc
.Op Fl dfhv
.Ar :group
.Ar
.Nm
.Oo
.Fl R
.Op Fl H | Fl L | Fl P
.Oc
.Op Fl dfhv
.Fl Fl reference=rfile
.Ar
.Sh DESCRIPTION
.Nm
sets the user ID and/or the group ID of the specified files.
Symbolic links named by arguments are silently left unchanged unless
.Fl h
is used.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl H
If the
.Fl R
option is specified, symbolic links on the command line are followed.
(Symbolic links encountered in the tree traversal are not followed.)
.It Fl L
If the
.Fl R
option is specified, all symbolic links are followed.
.It Fl P
If the
.Fl R
option is specified, no symbolic links are followed.
.It Fl R
Change the user ID and/or the group ID for the file hierarchies rooted
in the files instead of just the files themselves.
.It Fl d
Do not attempt to update a file's owner or group or its set-user-id
and set-group-id bits if they are all already set to the desired values.
.It Fl f
Do not report any failure to change file owner or group, nor modify
the exit status to reflect such failures.
.It Fl h
If
.Ar file
is a symbolic link, the owner and/or group of the link is changed.
.It Fl v
Cause
.Nm
to be verbose, showing files as they are processed.
.El
.Pp
The
.Fl H ,
.Fl L
and
.Fl P
options are ignored unless the
.Fl R
option is specified.
In addition, these options override each other and the
command's actions are determined by the last one specified.
The default is as if the
.Fl P
option had been specified.
.Pp
The
.Fl L
option cannot be used together with the
.Fl h
option.
.Pp
The
.Ar owner
and
.Ar group
operands are both optional, however, one must be specified; alternatively,
both the owner and group may be specified using a reference
.Ar rfile
specified using the
.Fl Fl reference
argument.
If the
.Ar group
operand is specified, it must be preceded by a colon (``:'') character.
.Pp
The
.Ar owner
may be either a user name or a numeric user ID.
The
.Ar group
may be either a group name or a numeric group ID.
Since it is valid to have a user or group name that is numeric (and
does not have the numeric ID that matches its name) the name lookup
is always done first.
Preceding an ID with a ``#'' character will force it to be taken
as a number.
.Pp
The ownership of a file may only be altered by a super-user for
obvious security reasons.
.Pp
Unless invoked by the super-user,
.Nm
clears the set-user-id and set-group-id bits on a file to prevent
accidental or mischievous creation of set-user-id and set-group-id
programs.
.Sh EXIT STATUS
.Ex -std chown
.Sh COMPATIBILITY
Previous versions of the
.Nm
utility used the dot (``.'') character to distinguish the group name.
This has been changed to be a colon (``:'') character so that user and
group names may contain the dot character.
.Sh SEE ALSO
.Xr chflags 1 ,
.Xr chgrp 1 ,
.Xr find 1 ,
.Xr chown 2 ,
.Xr lchown 2 ,
.Xr fts 3 ,
.Xr symlink 7
.Sh STANDARDS
The
.Nm
command is expected to be POSIX 1003.2 compliant.
.Pp
The
.Fl v
and
.Fl d
options and the use of ``#'' to force a numeric lookup
are extensions to
.St -p1003.2 .
.Sh HISTORY
A
.Nm
utility appeared in
.At v1 .


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision: 1.2968 $>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------

Changes from NetBSD 10.0 to NetBSD 11.0:
	amdsmn(4): added trivial power management support [reinoud 20221218]
	amdccp(4): added trivial power management support [reinoud 20221218]
	libc: Update Unicode tables to Unicode 13.0.0. [wiz 20221220]
	libc: Update Unicode tables to Unicode 14.0.0. [wiz 20221220]
	ppp(4): Remove several non-functioning ioctls. [chs 20221221]
	libc: Update Unicode tables to Unicode 15.0.0. [wiz 20221221]
	dse(4): Added Dayna SCSI/Link ethernet driver. [nat 20221222]
	binutils: Updated to FSF binutils 2.39. [christos 20221224]
	sdhc(4): Add quirk for some Intel eMMC devices (like
		GIGABYTE MA10-ST0) to allow the driver to attach
		successfully. [msaitoh 20230105]
	cribbage(6): Add -y option to keep playing without asking
		(from Greywolf in PR 57164) [christos 20230106]
	x86: Enable the -z separate-code security feature by default in
		ld(1). Reduces the default max page size from 2M to
		4K on amd64. [christos 20230106]
	bind: Import version 9.16.37. [christos 20230125]
	xfwp(1): remove. [wiz 20230204]
	wscons(4): Ignore nonsense tab stops in vt100 emulation.
		[riastradh 20230223]
	nvmm(4): Filter CR4 bits on x86 SVM (AMD). In particular,
		prohibit PKE, Protection Key Enable, which requires
		some additional management of CPU state by nvmm. 
		[riastradh 20230223]
	ftp(1): Add option sslnoverify to control validation of SSL
		certificates. Certificate validation is now enabled
		by default. [mlelstv 20230225]
	ftp(1): Add netrc processing to fetch-mode (URL on command line) to
		enable options and autologin via netrc. [mlelstv 20230225]
	aarch64: Audit use of curcpu(9). [riastradh 20230225]
	xen: Memory barrier audit. Reduce unnecessary membars.
		[riastradh 20230225]
	x86: Mitigate MXCSR Configuration Dependent Timing CPU security
		flaw in kernel FPU use. [riastradh 20230225]
	kernel: Optimization: Omit needless membar when triggering softint
		in various ports. [riastradh 20230301]
	sh(1): Adjust tilde expansion to comply with upcoming POSIX
		Issue 8 requirements. [kre 20230306]
	libm: Add more long double transcendental functions (as aliases of
		double functions) from C99. [riastradh 20230313]
	tzdata: updated to 2023a (using the 2023agtz fork) [kre 20230323]
	tzdata: updated to 2023b (using the 2023bgtz fork) [kre 20230324]
	xen: Allow a PVH dom0 to use VGA as console. [bouyer 20230324]
	vioscsi(4), viornd(4), viomb(4): Mark VirtIO drivers as MPSAFE.
		[mlelstv 20230325]
	amiga: Fix bootblocks for Kickstart 3.2. [abs 20230325]
	libc: Add __atomic_is_lock_free to help stdatomic.h usage with
		GCC. [riastradh 20230330]
	tzdata: updated to 2023c (using the 2023cgtz fork) [kre 20230331]
	tprof(8): Add Intel Comet Lake support. [msaitoh 20230410]
	vacation(1): Check 'Auto-Submitted:' header in addition to
		'Precedence:', and set 'Precedence:' in addition to
		'Auto-Submitted:'.  [hauke 20230411]
	ichsmb(4), tco(4): Add support for TCO on newer Intel chipsets.
		[riastradh 20230412]
	audiorecord(1): Add support for recording 24-bit wav files.
		[mlelstv 20230415]
	lua: Import version 5.4.4.
		Bumped shlib major because of incompatible API/ABI changes.
		[nikita 20230416]
	evbarm: Install a boot.cfg file for the EFI bootloader by default.
		[brook 20230419]
	dhcpcd: Import version 10.0.1. [roy 20230421]
	evbarm: Enable rkv1crypto driver on Rock64. [mrg 20230424]
	worms(6): Add a little bit of optional colour, code and
		ideas from RVP. [kre 20230426]
	amiga: Add driver for the MNT ZZ9000 Zorro board - submitted by
		Alain Runa. [phx 20230503]
	evbarm: Experimental IMX6SX support - tested on the UDOO Neo Full board.
		[bouyer 20230504]
	chown(8): Add a -d flag to avoid changing a file's owner/group to the
		current value.  This avoids some unnecessary operations on the
		file. [pgoyette 20230504]
	ftp(1): Add timeout for SSL connection setup, defaulting to 60 seconds.
		[lukem 20230505]
	riscv: Add RISC-V support that works on QEMU with a single hart.
		[skrll 20230507]
	riscv: Add Google Goldfish RTC driver. [skrll 20230508]
	OpenSSL: Imported 3.0.8. [christos 20230510]
	ctfmerge(1): Add error handling to pthread APIs.
		Fix semaphores for macOS host tool.  [lukem 20230523]
	share/mk: Deprecate non-functional SHAREDSTRINGS. [lukem 20230524]
	mkstr(1), xstr(1): Obsolete unused PDP-11-era tools. [lukem 20230524]


sbin/chown/chown.c
/*	$NetBSD: chown.c,v 1.12 2023/05/04 18:34:55 christos Exp $	*/

/*
 * Copyright (c) 1988, 1993, 1994, 2003
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#ifndef lint
__COPYRIGHT("@(#) Copyright (c) 1988, 1993, 1994, 2003\
 The Regents of the University of California.  All rights reserved.");
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)chown.c	8.8 (Berkeley) 4/4/94";
#else
__RCSID("$NetBSD: chown.c,v 1.12 2023/05/04 18:34:55 christos Exp $");
#endif
#endif /* not lint */

#include <sys/types.h>
#include <sys/stat.h>

#include <ctype.h>
#include <dirent.h>
#include <err.h>
#include <errno.h>
#include <locale.h>
#include <fts.h>
#include <grp.h>
#include <pwd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <getopt.h>

static void	a_gid(const char *);
static void	a_uid(const char *);
static id_t	id(const char *, const char *);
__dead static void	usage(void);

static uid_t uid;
static gid_t gid;
static int ischown;
static const char *myname;

struct option chown_longopts[] = {
	{ "reference",		required_argument,	0,
						1 },
	{ NULL,			0,			0,
						0 },
};

int
main(int argc, char **argv)
{
	FTS *ftsp;
	FTSENT *p;
	int Hflag, Lflag, Rflag, ch, fflag, fts_options, hflag, rval, vflag, dflag;
	char *cp, *reference;
	int (*change_owner)(const char *, uid_t, gid_t);

	setprogname(*argv);

	(void)setlocale(LC_ALL, "");

	myname = getprogname();
	ischown = (myname[2] == 'o');
	reference = NULL;

	Hflag = Lflag = Rflag = fflag = hflag = vflag = dflag = 0;
	while ((ch = getopt_long(argc, argv, "HLPRdfhv",
	    chown_longopts, NULL)) != -1)
		switch (ch) {
		case 1:
			reference = optarg;
			break;
		case 'd':
			dflag = 1;
			break;
		case 'H':
			Hflag = 1;
			Lflag = 0;
			break;
		case 'L':
			Lflag = 1;
			Hflag = 0;
			break;
		case 'P':
			Hflag = Lflag = 0;
			break;
		case 'R':
			Rflag = 1;
			break;
		case 'f':
			fflag = 1;
			break;
		case 'h':
			/*
			 * In System V the -h option causes chown/chgrp to
			 * change the owner/group of the symbolic link.
			 * 4.4BSD's symbolic links didn't have owners/groups,
			 * so it was an undocumented noop.
			 * In NetBSD 1.3, lchown(2) is introduced.
			 */
			hflag = 1;
			break;
		case 'v':
			vflag = 1;
			break;
		case '?':
		default:
			usage();
		}
	argv += optind;
	argc -= optind;

	if (argc == 0 || (argc == 1 && reference == NULL))
		usage();

	fts_options = FTS_PHYSICAL;
	if (Rflag) {
		if (Hflag)
			fts_options |= FTS_COMFOLLOW;
		if (Lflag) {
			if (hflag)
				errx(EXIT_FAILURE,
				    "the -L and -h options "
				    "may not be specified together.");
			fts_options &= ~FTS_PHYSICAL;
			fts_options |= FTS_LOGICAL;
		}
	} else if (!hflag)
		fts_options |= FTS_COMFOLLOW;

	uid = (uid_t)-1;
	gid = (gid_t)-1;
	if (reference == NULL) {
		if (ischown) {
			if ((cp = strchr(*argv, ':')) != NULL) {
				*cp++ = '\0';
				a_gid(cp);
			}
#ifdef SUPPORT_DOT
			else if ((cp = strrchr(*argv, '.')) != NULL) {
				if (uid_from_user(*argv, &uid) == -1) {
					*cp++ = '\0';
					a_gid(cp);
				}
			}
#endif
			a_uid(*argv);
		} else
			a_gid(*argv);
		argv++;
	} else {
		struct stat st;

		if (stat(reference, &st) == -1)
			err(EXIT_FAILURE, "Cannot stat `%s'", reference);
		if (ischown)
			uid = st.st_uid;
		gid = st.st_gid;
	}

	if ((ftsp = fts_open(argv, fts_options, NULL)) == NULL)
		err(EXIT_FAILURE, "fts_open");

	for (rval = EXIT_SUCCESS; (p = fts_read(ftsp)) != NULL;) {
		change_owner = chown;
		switch (p->fts_info) {
		case FTS_D:
			if (!Rflag)		/* Change it at FTS_DP. */
				fts_set(ftsp, p, FTS_SKIP);
			continue;
		case FTS_DNR:			/* Warn, chown, continue. */
			warnx("%s: %s", p->fts_path, strerror(p->fts_errno));
			rval = EXIT_FAILURE;
			break;
		case FTS_ERR:			/* Warn, continue. */
		case FTS_NS:
			warnx("%s: %s", p->fts_path, strerror(p->fts_errno));
			rval = EXIT_FAILURE;
			continue;
		case FTS_SL:			/* Ignore unless -h. */
			/*
			 * All symlinks we found while doing a physical
			 * walk end up here.
			 */
			if (!hflag)
				continue;
			/*
			 * Note that if we follow a symlink, fts_info is
			 * not FTS_SL but FTS_F or whatever.  And we should
			 * use lchown only for FTS_SL and should use chown
			 * for others.
			 */
			change_owner = lchown;
			break;
		case FTS_SLNONE:		/* Ignore. */
			/*
			 * The only symlinks that end up here are ones that
			 * don't point to anything.  Note that if we are
			 * doing a physical walk, we never reach here unless
			 * we asked to follow explicitly.
			 */
			continue;
		default:
			break;
		}

		/*
		 * If dflag was set, and the owner and group are already
		 * set to the right values and the set-user-id and
		 * set-group-id bits are both already clear, skip any
		 * attempt to update.
		 */
		if (dflag &&
		    (uid == (uid_t)-1 || p->fts_statp->st_uid == uid) &&
		    (gid == (gid_t)-1 || p->fts_statp->st_gid == gid) &&
		    (p->fts_statp->st_mode & 07000) == 0)
			continue;

		if ((*change_owner)(p->fts_accpath, uid, gid) && !fflag) {
			warn("%s", p->fts_path);
			rval = EXIT_FAILURE;
		} else {
			if (vflag)
				printf("%s\n", p->fts_path);
		}
	}
	if (errno)
		err(EXIT_FAILURE, "fts_read");
	exit(rval);
	/* NOTREACHED */
}

static void
a_gid(const char *s)
{
	struct group *gr;

	if (*s == '\0')			/* Argument was "uid[:.]". */
		return;
	gr = *s == '#' ? NULL : getgrnam(s);
	if (gr == NULL)
		gid = id(s, "group");
	else
		gid = gr->gr_gid;
	return;
}

static void
a_uid(const char *s)
{
	if (*s == '\0')			/* Argument was "[:.]gid". */
		return;
	if (*s == '#' || uid_from_user(s, &uid) == -1) {
		uid = id(s, "user");
	}
	return;
}

static id_t
id(const char *name, const char *type)
{
	id_t val;
	char *ep;

	errno = 0;
	if (*name == '#')
		name++;
	val = (id_t)strtoul(name, &ep, 10);
	if (errno)
		err(EXIT_FAILURE, "%s", name);
	if (*ep != '\0')
		errx(EXIT_FAILURE, "%s: invalid %s name", name, type);
	return (val);
}

static void
usage(void)
{

	(void)fprintf(stderr,
	    "Usage: %s [-R [-H | -L | -P]] [-dfhv] %s file ...\n"
	    "\t%s [-R [-H | -L | -P]] [-dfhv] --reference=rfile file ...\n",
	    myname, ischown ? "owner:group|owner|:group" : "group",
	    myname);
	exit(EXIT_FAILURE);
}
